Source/NonlinearSolvers/CurlCurlMLMGPC.H

```diff
@@ -0,0 +1,355 @@
+    if (m_use_gmres) {
+        Print() << PreconditionerTypes::curl_curl_mlmg
+                << " use PC for GMRES:     "
+                << (m_use_gmres_pc?"true":"false") << "\n";
+    }
+}
+
+template <class T, class Ops>
+void CurlCurlMLMGPC<T,Ops>::readParameters()
+{
+    const amrex::ParmParse pp(PreconditionerTypes::curl_curl_mlmg);
+    pp.query("verbose", m_verbose);
+    pp.query("bottom_verbose", m_bottom_verbose);
+    pp.query("max_iter", m_max_iter);
+    pp.query("agglomeration", m_agglomeration);
+    pp.query("consolidation", m_consolidation);
+    pp.query("max_coarsening_level", m_max_coarsening_level);
+    pp.query("absolute_tolerance",  m_atol);
+    pp.query("relative_tolerance",  m_rtol);
+    pp.query("use_gmres",  m_use_gmres);
+    pp.query("use_gmres_pc",  m_use_gmres_pc);
+}
+
+template <class T, class Ops>
+void CurlCurlMLMGPC<T,Ops>::Define ( const T& a_U,
+                                     Ops* const a_ops )
+{
+    using namespace amrex;
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        !IsDefined(),
+        "CurlCurlMLMGPC::Define() called on defined object" );
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        (a_ops != nullptr),
+        "CurlCurlMLMGPC::Define(): a_ops is nullptr" );
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        a_U.getArrayVecType()==warpx::fields::FieldType::Efield_fp,
+        "CurlCurlMLMGPC::Define() must be called with Efield_fp type");
+
+    m_ops = a_ops;
+    // read preconditioner parameters
+    readParameters();
+
+// currently not implemented in 1D
+#ifdef WARPX_DIM_1D_Z
+    WARPX_ABORT_WITH_MESSAGE("CurlCurlMLMGPC not yet implemented for 1D");
+#else
+    // create info object for curl-curl op
+    m_info = std::make_unique<LPInfo>();
+    m_info->setAgglomeration(m_agglomeration);
+    m_info->setConsolidation(m_consolidation);
+    m_info->setMaxCoarseningLevel(m_max_coarsening_level);
+
+    // Get data vectors from a_U
+    auto& u_mfarrvec = a_U.getArrayVec();
+
+    // Set number of AMR levels and create geometry, grids, and
+    // distribution mapping vectors.
+    m_num_amr_levels = m_ops->numAMRLevels();
+    m_geom.resize(m_num_amr_levels);
+    m_grids.resize(m_num_amr_levels);
+    m_dmap.resize(m_num_amr_levels);
+    for (int n = 0; n < m_num_amr_levels; n++) {
+        m_geom[n] = m_ops->GetGeometry(n);
+        m_dmap[n] = u_mfarrvec[n][0]->DistributionMap();
+
+        BoxArray ba = u_mfarrvec[n][0]->boxArray();
+        m_grids[n] = ba.enclosedCells();
+    }
+
+    // Construct the curl-curl linear operator and set its BCs
+    m_curl_curl = std::make_unique<MLCurlCurl>(m_geom, m_grids, m_dmap, *m_info);
+    m_curl_curl->setDomainBC(m_ops->GetLinOpBCLo(), m_ops->GetLinOpBCHi());
+
+    // Dummy value for alpha and beta to avoid abort due to degenerate matrix by MLMG solver
+    m_curl_curl->setScalars(1.0, 1.0);
+
+    // Construct the MLMG solver
+    m_solver = std::make_unique<MLMGT<MFArr>>(*m_curl_curl);
+    m_solver->setMaxIter(m_max_iter);
```

Comment:
```suggestion
    m_solver->setMaxIter(m_max_iter);
    m_solver->setFixedIter(m_max_iter);
```
If we don't call setFixedIter and MLMG fails to converge to the tolerance target within m_max_iter iterations, it would abort. Since this is used as a preconditioner, not a full solver, we don't really want it to abort.